### ISSUE TITLE
opx-ethtool set speed is not reflected or taking affect

### ISSUE THREAD

**Issue by Ragsboss** [open-switch/opx-nas-interface#22]
See below even after setting speed it still shows as zero

```
root@cumulus:/var/log/aos# opx-ethtool e101-003-0
Settings for e101-003-0:
    Channel ID:   0
    Transceiver Status: Disable
    Media Type: SFPPLUS 10GBASE CU1M
    Part Number: 599700002
    Serial Number: APF12230025JUC
    Qualified: Yes
    Administrative State: DOWN
    Operational State: DOWN
    Supported Speed (in Mbps):  [1000, 10000]
    Auto Negotiation : off
    Speed   : 0
    Duplex   : full
root@cumulus:/var/log/aos# opx-ethtool -s e101-003-0 speed 10000    
 speed 10000
root@cumulus:/var/log/aos# opx-ethtool e101-003-0
Settings for e101-003-0:
    Channel ID:   0
    Transceiver Status: Disable
    Media Type: SFPPLUS 10GBASE CU1M
    Part Number: 599700002
    Serial Number: APF12230025JUC
    Qualified: Yes
    Administrative State: DOWN
    Operational State: DOWN
    Supported Speed (in Mbps):  [1000, 10000]
    Auto Negotiation : off
    Speed   : 0
    Duplex   : full
```

Info about the above port
```
root@cumulus:/var/log/aos# opx-ethtool -e e101-003-0
Show media info for e101-003-0
Key: 2.28.1835221.1835077.1835078.1835079.
base-pas/media/rate-identifier = 0
base-pas/media/oper-status = 0
base-pas/media/category = 2
base-pas/media/voltage-state = 1
base-pas/media/bias-low-warning-threshold = 0.0
base-pas/media/vendor-pn = 599700002
base-pas/media/current-temperature = 0.0
base-pas/media/insertion-cnt = 1
base-pas/media/voltage-low-warning-threshold = 0.0
base-pas/media/cc_ext = 242
base-pas/media/length-om2 = 0
base-pas/media/length-om3 = 0
base-pas/media/rx-power-low-alarm-threshold = 0.0
base-pas/media/length-om1 = 0
base-pas/media/vendor-id = 415048
base-pas/media/media-category/sfp-plus/br-max = 0
base-pas/media/connector = 33
base-pas/media/ext-transceiver = 0
base-pas/media/media-category/sfp-plus/br-min = 0
base-pas/media/encoding = 0
base-pas/media/tx-power-high-warning-threshold = 0.0
base-pas/media/vendor-name = Amphenol
base-pas/media/rx-power-low-warning-threshold = 0.0
base-pas/media/slot = 1
base-pas/media/port = 3
base-pas/media/vendor-rev = 4300
base-pas/media/slot = 1
base-pas/media/port = 3
base-pas/media/tx-power-low-alarm-threshold = 0.0
base-pas/media/bias-low-alarm-threshold = 0.0
base-pas/media/capability = 4
base-pas/media/media-category/sfp-plus/sff-8472-compliance = 0
base-pas/media/diag-mon-type = 0
base-pas/media/temp-state = 2
base-pas/media/type = 13
base-pas/media/media-category/qsfp-plus/wavelength-tolerance = 0
base-pas/media/voltage-high-warning-threshold = 0.0
base-pas/media/ext-identifier = 4
base-pas/media/temp-low-warning-threshold = 0.0
base-pas/media/qualified = 1
base-pas/media/temp-high-alarm-threshold = 0.0
base-pas/media/length-sfm = 0
base-pas/media/rate-select-state = 0
base-pas/media/rx-power-measurement-type = 0
base-pas/media/wavelength = 256
base-pas/media/cc_base = 245
base-pas/media/temp-low-alarm-threshold = 0.0
base-pas/media/tx-power-low-warning-threshold = 0.0
base-pas/media/insertion-timestamp = 140368180288976
base-pas/media/current-voltage = 0.0
base-pas/media/bias-high-alarm-threshold = 0.0
base-pas/media/high-power-mode = 0
base-pas/media/br-nominal = 103
base-pas/media/options = 0
base-pas/media/support-status = 0
base-pas/media/date-code = 3132303631300000
base-pas/media/present = 1
base-pas/media/transceiver = 0000000000100000
base-pas/media/length-cable = 1
base-pas/media/rx-power-high-warning-threshold = 0.0
base-pas/media/voltage-high-alarm-threshold = 0.0
base-pas/media/identifier = 3
base-pas/media/voltage-low-alarm-threshold = 0.0
base-pas/media/length-sfm-km = 0
base-pas/media/rx-power-high-alarm-threshold = 0.0
base-pas/media/admin-status = 0
base-pas/media/serial-number = APF12230025JUC
base-pas/media/tx-power-high-alarm-threshold = 0.0
base-pas/media/temp-high-warning-threshold = 0.0
base-pas/media/bias-high-warning-threshold = 0.0
base-pas/media/enhanced-options = 0
base-pas/media/media-category/qsfp-plus/max-case-temp = 0
```

OPX version and switch info
```
root@cumulus:/var/log/aos# opx-show-version 
OS_NAME="OPX"
OS_VERSION="2.1.0"
PLATFORM=""
ARCHITECTURE="x86_64"
INTERNAL_BUILD_ID="OpenSwitch blueprint for Dell 1.0.0"
BUILD_VERSION="2.1.0(0)"
BUILD_DATE="2017-08-04T12:01:19-0700"
INSTALL_DATE="2017-10-11T04:42:24+0000"
SYSTEM_UPTIME= 1 week, 8 hours, 18 minutes
SYSTEM_STATE= running
UPGRADED_PACKAGES=no
ALTERED_PACKAGES=no
root@cumulus:/var/log/aos# opx-show-env 
Chassis
	Operating status:		Fail
	Fault type:			Configuration error
	Vendor name:		
	Service tag:		2DYMX42
	PPID:				CN0M68YC282985540009
	Platform name:			
	Product name:			S4048ON
	Hardware version:		A00
	Number of MAC addresses:	256
	Base MAC address:		34:17:eb:f4:f2:c4
Power supplies
	Slot 1
		Present:		Yes
		Operating status:	Up
		Fault type:		OK
		Vendor name:		
		Service tag:		AEIOU##
		PPID:			CN00YNWT2829855F0085
		Platform name:			
		Product name:		
		Hardware version:		A00
		Input:			AC
		Fan airflow:		Reverse
	Slot 2
		Present:		No
Fan trays
	Slot 1
		Present:		Yes
		Operating status:	Up
		Fault type:		OK
		Vendor name:		
		Service tag:		AEIOU##
		PPID:			CN03HC752829855C0021
		Platform name:			
		Product name:		
		Hardware version:		A00
		Fan airflow:		Reverse
	Slot 2
		Present:		Yes
		Operating status:	Up
		Fault type:		OK
		Vendor name:		
		Service tag:		AEIOU##
		PPID:			CN03HC752829855F0271
		Platform name:			
		Product name:		
		Hardware version:		A00
		Fan airflow:		Reverse
	Slot 3
		Present:		Yes
		Operating status:	Up
		Fault type:		OK
		Vendor name:		
		Service tag:		AEIOU##
		PPID:			CN03HC752829855F0293
		Platform name:			
		Product name:		
		Hardware version:		A00
		Fan airflow:		Reverse
Fans
	Fan 1, PSU slot 1
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		10768
		Speed (%):		59
	Fan 1, Fan tray slot 1
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		10082
		Speed (%):		53
	Fan 2, Fan tray slot 1
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		10056
		Speed (%):		52
	Fan 1, Fan tray slot 2
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		9867
		Speed (%):		51
	Fan 2, Fan tray slot 2
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		9892
		Speed (%):		52
	Fan 1, Fan tray slot 3
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		10069
		Speed (%):		52
	Fan 2, Fan tray slot 3
		Operating status:	Up
		Fault type:		OK
		Speed (RPM):		10160
		Speed (%):		53
Temperature sensors
	Sensor CPU board sensor, Card slot 1
		Operating status:		Up
		Fault type:			OK
		Temperature (degrees C):	23
	Sensor NPU board sensor, Card slot 1
		Operating status:		Up
		Fault type:			OK
		Temperature (degrees C):	28
	Sensor system-NIC board sensor 1, Card slot 1
		Operating status:		Up
		Fault type:			OK
		Temperature (degrees C):	26
	Sensor system-NIC board sensor 2, Card slot 1
		Operating status:		Up
		Fault type:			OK
		Temperature (degrees C):	25
	Sensor NPU temp sensor, Card slot 1
		Operating status:		Up
		Fault type:			OK
		Temperature (degrees C):	42
```

fyi: @deltasierra

**Comment by GarrickHe** [comment]
Hello @Ragsboss @deltasierra,
I'm working on this issue. Will keep you posted on my progress.

Thank you,
Garrick

**Comment by GarrickHe** [comment]
Hello @Ragsboss @deltasierra,
I found the issue and determined that it is expected behavior. The problem is your link is not in operational state so the speed will always be 0 since there is no way to determine what the actual speed will be.

Thank you,
Garrick

**Comment by GarrickHe** [comment]
Hello @Ragsboss @deltasierra,
Fix has been pushed into opx-nas-l2:

Fix: traceback seen when using 'opx-config-switch show'.
Change-Id: I0e6de836fcfbcacff9f3a772d2851f9392b284bb
Signed-off-by: Garrick He <garrick_he@dell.com>

Thank you,
Garrick

**Comment by Ragsboss** [comment]
I'm sorry I didn't see your response for some reason. What do you mean by 'link is not in operational state' - I'm having this problem again? This switch was working fine and we re-imaged the neighboring switch and now again I'm seeing the link is down on OPX and speed setting doesn't take affect. The administrative state is UP but operation state is DOWN on the link.

```
root@s6000-r2r7-u28-spine2:/var/log/aos# opx-ethtool -s e101-008-0 speed 40000 duplex full
 speed 40000 duplex full
root@s6000-r2r7-u28-spine2:/var/log/aos# opx-ethtool e101-008-0
Settings for e101-008-0:
    Channel ID:   0
    Transceiver Status: Disable
    Channel ID:   1
    Transceiver Status: Disable
    Channel ID:   2
    Transceiver Status: Disable
    Channel ID:   3
    Transceiver Status: Disable
    Media Type: QSFP 40GBASE CR4
    Part Number: 599690001
    Serial Number: APF12210014RWJ
    Qualified: Yes
    Administrative State: UP
    Operational State: DOWN
    Supported Speed (in Mbps):  [1000, 10000, 40000]
    Auto Negotiation : on
    Speed   : 0
    Duplex   : full
root@s6000-r2r7-u28-spine2:/var/log/aos# ip link show dev e101-008-0
5: e101-008-0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether ec:f4:bb:fc:d7:82 brd ff:ff:ff:ff:ff:ff
    alias facing_dual-l3-compute-1-leaf2:swp49
```

Appreciate if we can get to the bottom of this..

**Comment by GarrickHe** [comment]
Hello @Ragsboss,

This usually means the port is enabled and powered on properly, but the line protocol has an issue. It maybe something wrong with the other side. Possible reasons:

1. It's not attached to anything.
2. It can't negotiate a speed that you want.

Can you please check the peer to make sure everything is proper, if you haven't already done so.

Hope this helps,
Garrick

**Comment by Ragsboss** [comment]
@GarrickHe you are right! The other side (a Cumulus 3.3.2) has auto-negation off (I'm still figuring how/why this happened) but once I set it on - things are working.

One difference I observed though is `ethtool` on the other side does reflect the set speed but `opx-ethtool` does NOT reflect the set speed - as in, it does not show that the configuration itself is accepted, thus sometimes confusing and misleading the diagnosis..

**Comment by GarrickHe** [comment]
@Ragsboss how did you create this error? I just want to know the steps to reproduce. It maybe a bug. Here is my understanding of your steps:

1. use 'opx-ethtool' to set the port speed on the box running OPX, while the peer has auto-neg turned off. The configuration fails with since there is a line protocol issue.
2. turn on auto-neg on the peer and everything works but the opx side doesn't have the configured speed set in step #1?

Is this correct?

thanks,
Garrick

**Comment by Ragsboss** [comment]
No. After step 2 opx side shows correct speed

Sent from my iPhone
Email maybe terse

On Dec 5, 2017, at 1:54 PM, Garrick He <notifications@github.com<mailto:notifications@github.com>> wrote:


@Ragsboss<https://github.com/ragsboss> how did you create this error? I just want to know the steps to reproduce. It maybe a bug. Here is my understanding of your steps:

  1.  use 'opx-ethtool' to set the port speed on the box running OPX, while the peer has auto-neg turned off. The configuration fails with since there is a line protocol issue.
  2.  turn on auto-neg on the peer and everything works but the opx side doesn't have the configured speed set in step #1<https://github.com/open-switch/opx-nas-interface/pull/1>?

Is this correct?

thanks,
Garrick

**Comment by GarrickHe** [comment]
Hello @Ragsboss ,

Sorry for the delayed response. I'm still a bit confused about your issue. First you say the opx-side (using opx-ethtool) doesn't reflect the correct speed. But the previous comment you said after step 2 opx-side shows the correct speed. Can you please clarify the steps you took and the issue you see. 

From your previous comment:

"One difference I observed though is ethtool on the other side does reflect the set speed but opx-ethtool does NOT reflect the set speed - as in, it does not show that the configuration itself is accepted, thus sometimes confusing and misleading the diagnosis.."

What (incorrect) speed does the opx-side show you? I'm guessing your issue is before enabling auto-neg on the peer (Cumulus) the set speed on the opx side was not accepted. But once you turned on auto-neg on the peer and things worked, did you set the speed on the opx side again? Since the initial configuration on the OPX side wasn't accepted, it will still be using auto-neg.

**Comment by Ragsboss** [comment]
I'll ping back here if we see the issue again and I'll try to be more clear. Thank you for looking into it.
